Request Buffer

vermapratyush opened the pull request:
Adds a buffer on requests before throwing ErrMaxConcurrency.
The Timeout is still adhered to, timer starts when the request is submitted to hystrix-go, not when execution starts. This basically implements MaxQueueSize as present in the [Netflix's Hystrix](https://github.com/Netflix/Hystrix/wiki/Configuration#maxqueuesize)

The default value of `MaxQueueSize is 50 (5 * MaxConcurrency)`, although can be overridden when initialising circuit.

In addition to the request buffer, the PR includes a different way to solving for #67 . It uses channels instead of `sync.Once`. Also some go-lint fixes.

We have been running this in production at Grab for a while, and there seems to be no side effects.

tonyghita commented on hystrix/eventstream.go:
```diff
@@ -162,13 +165,13 @@ func (sh *StreamHandler) publishThreadPools(pool *executorPool) error {
 		return err
 	}
-	err = sh.writeToRequests(eventBytes)
+	_ = sh.writeToRequests(eventBytes)
```
Should probably `return sh.writeToRequests(eventBytes)` instead of ignoring the error.

vermapratyush replied:
fixed

tonyghita commented on .gitignore:
```diff
@@ -1 +1,3 @@
-.vagrant
+*.iml
```
The IDE-specific ignore lines probably belong in  your global .gitignore: https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

vermapratyush replied:
fixed

tonyghita commented on README.md:
````diff
@@ -87,9 +87,10 @@ During application boot, you can call ```hystrix.ConfigureCommand()``` to tweak
+	Timeout:                     1000,
+	MaxConcurrentRequests:       100,
+	ErrorPercentThreshold:       25,
+	QueueSizeRejectionThreshold: 100,
````
How would you pick a good value for `QueueSizeRejectionThreshold`?

vermapratyush replied:
The default value of `QueueSizeRejectionThreshold` is currently equal to the `MaxConcurrentRequests`. This should take care of request spike which is 2x the usual.
2x seems to be a decent number, although in some of the use-cases I have seen 3x-4x as well (in my workplace).
Netflix also use a number equal to `MaxConcurrentRequests`, although I am open to suggestion for a better default value.

tamccall commented on hystrix/metric_collector/metric_collector.go:
```diff
@@ -45,6 +45,8 @@ func (m *metricCollectorRegistry) Register(initMetricCollector func(string) Metr
 	// IncrementAttempts increments the number of updates.
 	IncrementAttempts()
+	// IncrementQueueSize increments the number of elements in the queue.
+	IncrementQueueSize()
```
This would break any external implementations of this interface. Should probably start managing releases as suggested by https://github.com/afex/hystrix-go/issues/70 before merging this

afex commented on hystrix/settings.go:
```diff
@@ -80,12 +86,18 @@ func ConfigureCommand(name string, config CommandConfig) {
 		errorPercent = config.ErrorPercentThreshold
 	}
 
+	queueSizeRejectionThreshold := DefaultQueueSizeRejectionThreshold
```
your PR comment mentions that the default is the same as `MaxConcurrentRequests`, but in fact it is statically set to `50` here even if the user provides a different concurrency setting.

i like the idea of having the queue size (if unset) be equal to the concurrency setting, which would change the code here to remove `DefaultQueueSizeRejectionThreshold` and replace it with `max`

vermapratyush replied:
Fixed.

afex commented on hystrix/settings.go:
```diff
@@ -16,14 +16,18 @@ var (
 	// DefaultErrorPercentThreshold causes circuits to open once the rolling measure of errors exceeds this percent of requests
 	DefaultErrorPercentThreshold = 50
+	// DefaultQueueSizeRejectionThreshold reject requests when the queue size exceeds the given limit
+	DefaultQueueSizeRejectionThreshold = DefaultMaxConcurrent * 5
```
recommend removing this default based on other comment in this file

afex commented on hystrix/metrics.go:
```diff
@@ -95,6 +95,9 @@ func (m *metricExchange) IncrementMetrics(wg *sync.WaitGroup, collector metricCo
 	}
+	if update.Types[0] == "queued" {
+		collector.IncrementQueueSize()
+	}
```
as described in my comment in `hystrix.go`, this won't work to accurately track the event rate. a types list of `[ queued, failure, fallback-success ]` should apply all of:

```go
collector.IncrementQueueSize()
collector.IncrementFailures()
collector.IncrementAttempts()
collector.IncrementErrors()
collector.IncrementFallbackSuccesses()
```

afex commented on hystrix/metric_collector/metric_collector.go:
```diff
@@ -45,6 +45,8 @@ func (m *metricCollectorRegistry) Register(initMetricCollector func(string) Metr
 	// IncrementAttempts increments the number of updates.
 	IncrementAttempts()
+	// IncrementQueueSize increments the number of elements in the queue.
+	IncrementQueueSize()
```
"queue size/length" does not seem like an accurate name since this is better stated as "rate at which executions were queued", or "number of queued events over a time window"

vermapratyush replied:
I have changed the name to `IncrementQueuedItem`. I tried to keep it in sync with other function names like `IncrementAttempts`. Let me know if you think of some other function name to be more appropriate.

afex commented on hystrix/hystrix.go:
```diff
@@ -118,51 +97,89 @@ func Go(name string, run runFunc, fallback fallbackFunc) chan error {
-				returnTicket()
+			select {
+			case t := <-circuit.executorPool.WaitingTicket:
+				cmd.reportEvent("queued")
```
adding a new event type is problematic here. i agree that we should track the rate at which executions are queued, but there is a current assumption being made about the event list for an execution which no longer holds true here.

currently the `events []string` field of a command is assumed to contain data in a format of:

```
[ success|failure|rejected|short-circuit|timeout, fallback-success|fallback-failure ]
```

for example, an events slice containing `[ failure, fallback-success ]` indicates the execution failed but the fallback did not.  changing this to `[ queued, failure, fallback-success ]` makes sense (execution was queued, then failed, then fell back successfully) but other parts of the code assume the first element in the list indicates the run function's result.  this changes that and will break stats reporting as well as closing a circuit after a success.

in order to add this queued event, you'll need to change `CircuitBreaker.ReportEvent` and `metricExchange.IncrementMetrics` to account for this.

vermapratyush replied:
Nice catch, I have incorporated the change by using a map structure to pass around the events.

vermapratyush commented:
@afex Thanks for the review. I have made the required changes in the PR.

In order to address the regression for buffer queue, I have added comment in `loadtest/README.md`. I have just mentioned that increasing the concurrency in bench tool should validate the buffer implementation. Let me know if you think a much more comprehensive test suite is required.

Changes:
1. Refactor function name `IncrementQueuedItem`
2. Fix bug in events being reported.
3. Incorporate changes related to queue size.
4. Added comment to run regression for queue length.
